pkg/kubelet/nodestatus/setters.go

```diff
@@ -197,36 +198,65 @@ func NodeAddress(nodeIP net.IP, // typically Kubelet.nodeIP
 			// For steps 3 and 4, IPv4 addresses are preferred to IPv6 addresses
 			// unless nodeIP is "::", in which case it is reversed.
 			if nodeIPSpecified {
-				ipAddr = nodeIP
+				ipAddr = []net.IP{nodeIP}
 			} else if addr := net.ParseIP(hostname); addr != nil {
-				ipAddr = addr
+				ipAddr = []net.IP{addr}
 			} else {
 				var addrs []net.IP
 				addrs, _ = net.LookupIP(node.Name)
 				for _, addr := range addrs {
 					if err = validateNodeIPFunc(addr); err == nil {
-						if isPreferredIPFamily(addr) {
-							ipAddr = addr
-							break
-						} else if ipAddr == nil {
-							ipAddr = addr
+						if utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) {
+							ipAddr = append(ipAddr, addr)
+						} else {
+							if isPreferredIPFamily(addr) {
+								ipAddr = []net.IP{addr}
+								break
+							} else if len(ipAddr) < 1 {
+								ipAddr = []net.IP{addr}
+							}
 						}
 					}
 				}
 
-				if ipAddr == nil {
-					ipAddr, err = utilnet.ResolveBindAddress(nodeIP)
+				// The `net.LookupIP` above can be short-circuited by the local
+				// /etc/hosts file, so we try again using only DNS
```

Comment:
Why? If the system is configured so that `/etc/hosts` overrides DNS, and `/etc/hosts` has an entry for the hostname, then we should *respect* that, not thwart it.

Reply:
We are respecting it.  The issue is that the hosts file can sometimes point to the loopback interface, which not useful to the kubelet here.  This is why we are checkout to see if we have any validated IP addresses:  if the hosts file has a legitimate, non-loopback-interface map, we will honour it.

Reply:
I agree with Dan, nsswitch.conf should be the authority, entries in /etc/hosts should be a sysadmin responsibility

Reply:
Is this something that you personally need for the machines that you are running this patch on, or are you just trying to be super generic? If the former, why can't you fix the machines so that they don't map their own hostname to 127.0.0.1, which seems like an obviously incorrect configuration?

Reply:
There are a number of issues with `/etc/hosts`-based IP discovery.  I was resorting to the simplest solution.

It is fairly common for `/etc/hosts` to just point at a loopback interface, especially in situations where the machine's IP is not known in advance or when network connectivity is delayed, so that resolution to self will work.

The issue with dual-stack systems is even more problematic:  you may have SLAAC with variable prefixes, you may have privacy extensions, resulting an a dynamic succession of changing IP addresses.

Yes, sure:  you can say to just set `host.conf` or `nsswitch.conf` to always prefer DNS to local files, but those files do not exist on the systems I deploy, and the default priority is generally preferred, in any case:  you want to be able to override hosts locally.  The issue only arises when you are trying to discover your own IP address using this method.